internal/template-validator/tlsinfo/tlsinfo.go

```diff
@@ -45,100 +49,73 @@ func (ti *TLSInfo) initCryptoConfig() {
 	ti.sspTLSOptions.MinTLSVersion, _ = os.LookupEnv(TLSMinVersionEnvName)
 }
 
-func (ti *TLSInfo) initCerts() {
+func (ti *TLSInfo) initCerts() error {
+	if ti.stopWatch != nil {
+		return fmt.Errorf("certificate watcher was already initialized")
+	}
+
 	directory := ti.CertsDirectory
-	filesChanged, watcherCloser, err := watchDirectory(directory)
+	ti.updateCertificates(directory)
+
+	ti.watch = filewatch.New()
+	err := ti.watch.Add(directory, func() {
+		ti.updateCertificates(directory)
+	})
 	if err != nil {
-		panic(err)
+		return fmt.Errorf("error adding directory to watch: %w", err)
 	}
 
-	ti.stopCertReload = make(chan struct{})
-	notify(filesChanged)
-
+	ti.stopWatch = make(chan struct{})
 	go func() {
-		defer watcherCloser.Close()
 		for {
 			select {
-			case <-filesChanged:
-				err := ti.updateCertificates(directory)
-				if err != nil {
-					go func() {
-						time.Sleep(retryInterval)
-						notify(filesChanged)
-					}()
-				}
-			case <-ti.stopCertReload:
+			case <-ti.stopWatch:
 				return
+			default:
 			}
-		}
-	}()
-}
-
-func (ti *TLSInfo) Clean() {
-	if ti.stopCertReload != nil {
-		close(ti.stopCertReload)
-	}
-}
-
-func watchDirectory(directory string) (chan struct{}, io.Closer, error) {
-	watcher, err := fsnotify.NewWatcher()
-	if err != nil {
-		logger.Log.Error(err, "Failed to create an inotify watcher")
-		return nil, nil, err
-	}
 
-	err = watcher.Add(directory)
-	if err != nil {
-		watcher.Close()
-		logger.Log.Error(err, "Failed to establish a watch", "directory", directory)
-		return nil, nil, err
-	}
+			watchErr := ti.watch.Run(ti.stopWatch)
+			if watchErr == nil {
+				return
+			}
+			// Log error and restart watch
+			logger.Log.Info("failed watching files", "error", watchErr)
 
-	filesChanged := make(chan struct{}, 1)
-	go func() {
-		for {
 			select {
-			case event, ok := <-watcher.Events:
-				if !ok {
-					return
-				}
-				if event.Op != fsnotify.Chmod {
-					notify(filesChanged)
-				}
-			case err, ok := <-watcher.Errors:
-				if !ok {
-					return
-				}
-				logger.Log.Error(err, "An error occurred while watching", "directory", directory)
+			case <-ti.stopWatch:
+				return
+			case <-time.After(time.Second):
 			}
 		}
 	}()
 
-	return filesChanged, watcher, nil
+	return nil
 }
 
-func notify(channel chan struct{}) {
-	select {
-	case channel <- struct{}{}:
-	default:
+func (ti *TLSInfo) Clean() {
+	if ti.stopWatch != nil {
+		close(ti.stopWatch)
+		ti.stopWatch = nil
 	}
 }
 
-func (ti *TLSInfo) updateCertificates(directory string) error {
+func (ti *TLSInfo) updateCertificates(directory string) {
 	cert, err := loadCertificates(directory)
+
+	ti.certLock.Lock()
```

Comment:
Can you help me to understand what this mutex does?

Reply:
It guards access to the fields `ti.cert` and `ti.certError`.
These fields are accessed from 2 goroutines:
- The file watcher goroutine, that updates these fields when the mouned secret is changed.
- The TLS listener goroutine, that reads the certificate when a new incoming connection is accepted.

Reply:
ACK